### ISSUE TITLE
Preview of MP4 Videos failing to generate with ffmpeg/avconf if larger than 5MB

### ISSUE THREAD

**Issue by sascha-schieferdecker** [nextcloud/server#1475]
### Steps to reproduce
1. Upload MP4 video larger than 5MB, with the moov atom at the end of the file, e.g. from an older android phone. About the moov atom: [https://www.adobe.com/devnet/video/articles/mp4_movie_atom.html](https://www.adobe.com/devnet/video/articles/mp4_movie_atom.html)

### Expected behaviour
Video Thumbnail should be generated, works with MPG files.

### Actual behaviour
avconf/ffmpeg fail because only the forst 5MB of the video are taken into account, see [https://github.com/nextcloud/server/blob/master/lib/private/Preview/Movie.php#L57](https://github.com/nextcloud/server/blob/master/lib/private/Preview/Movie.php#L57)

If  the moov atom is at the beginning of a video, the preview generation works. If the the moov atom is at end of the video, the generation fails. If it is a local file would it be possible to load it completely? Removing the limit in the stream_get_contents call generates the preview just fine. Or maybe would it be possible to streamline the video using something like `qt-faststart` first?

### Server configuration
**Operating system**: Debian 8

**Web server:** Apache 2.4

**Database:** MySQL

**PHP version:** PHP  7.0.11

**Nextcloud version:** Nextcloud 10.0 (stable)

**Updated from an older Nextcloud/ownCloud or fresh install:** No

**Where did you install Nextcloud from:** Homepage

Please forgive me if I don't post any other configuration details, IMHO they are not relevant to described behaviour.
``
</details>

**Comment by nickvergessen** [comment]
Hmm, disabling the switch based on storage sounds a bit weird to me.
I'd say this is fine as is, if the file is too big and does not have "progressive download," "fast start," "use streaming mode," or similar options, I guess that is simply what you end up with?

**Comment by sascha-schieferdecker** [comment]
Okay, if you define it as expected behaviour, this is what a user ends up with. I am a first time user of nextcloud (never used owncloud). I wondered why there was no preview of most mp4 videos, so I dug into the code. I think I'm not the only user wondering if this behaviour is an error. 

In my case I converted the videos to fast start and my problem is solved. But I think most users expect a working preview generation.

Maybe this is more of feature request, depending on how you see it.

**Comment by binarydad** [comment]
I'm noticing this now, but they're for SMB shares on my local network. Is there a way to force preview generation for videos over 5MB?

**Comment by sascha-schieferdecker** [comment]
AFAIK no, but you could convert the video with ffmpeg to "fast start", as [described here](https://gist.github.com/jaydenseric/220c785d6289bcfd7366).

**Comment by binarydad** [comment]
Thanks, but how can I tell nextcloud to use these options when generating previews?

**Comment by sascha-schieferdecker** [comment]
You can't AFAIK. Just run your videos through qt-faststart or ffmpeg in a batch  Sucked, but that's how I solved it.

**Comment by binarydad** [comment]
Alright, thank you for that. I'll take a look. 

Really bizarre... 5MB is nothing. This issue was racking my brain for months, as video previews for larger movies worked fine on local storage, but failed over SMB. At least I know why now...

**Comment by kaipee** [comment]
Is there a fix scheduled for this? I'm still failing to generate any video (MP4) thumbnails for files stored on 1Gbps SMB share

**Comment by liranbx** [comment]
I'm also a first time user running into this thing and had to dive into the code to figure it out (too bad this issue was not listed on my google searches).
Was there any progress on this? Is there a branch you're working on?

**Comment by binarydad** [comment]
So, the way I "solved" this was just bypass the code that is distinguishing between a local and a remote file. 

In `/var/www/nextcloud/lib/private/Preview/Movie.php`, in the `getThumbnail()` method, is the line `$absPath = $this->getLocalFile($file, 5242880); // only use the first 5MB`. This is where it's loading a file, but only capping it to 5MB. The `getLocalFile()` method is inside `ProviderV2.php` (same directory), which determines whether the file is local or not, eventually setting the `$useTempFile` boolean flag. If this is true, this is where it loads in an external file, but only 5MB. However, if it's told to load a local file, it will do so, and I've noticed this works fine for SMB files as well. 

My fix was to bypass the `$useTempFile` flag and force `getLocalFile()` to treat the "external" (SMB shared) file also as local, which allows FFMPEG to read the full file and generate a thumbnail. 

I've also tried simply increasing the max file size from 5MB to something much higher, but I feel the overhead of loading an external file the way it does was slower than simply bypassing it and treating it as a local file. I don't know much about PHP but I imagine going directly to the file system to get the file rather than PHP loading it into memory first would be quicker. And I have noticed that thumbnails for larger videos (several GB at least) generate thumbnails a little quicker. 

Here's my glamorous fix in PreviewV2.php: 🙃

![image](https://user-images.githubusercontent.com/1760607/107511350-59fb8900-6b73-11eb-852e-607942341509.png)

**Comment by kaipee** [comment]
@binarydad Just confirming that this 'fix' works great for me too. I have confirmed MP4 large videos are now very quickly generating thumbnails in the Nextcloud Android app, when they completely failed before.

BTW your last sentence should be updated, you've called the file PreviewV2.php instead of ProviderV2.php

**Comment by liranbx** [comment]
Sorry for the late comment.
From what I understand your solution works because you work with a high speed LAN SMB storage, sadly the same solution won't work for me as I use S3 object storage as a primary storage. It seems as there is no simple solution to creating mp4 previews on a truly remote storage except pulling the entire file without pre-processing the file, because the moov atom placement might be at the end of the file.

My thought is to create some flow application that will pre-process every video file upload to convert it either to webm format or to use qt-faststart to relocate the moov atom.

**Comment by binarydad** [comment]
Are you using the preview generator CRON job? You can increase the file 
limit from 5MB to something more reasonable for your case and schedule 
the preview generation to at least run at night time.

------ Original Message ------

**Comment by liranbx** [comment]
Problem is that with the large size of video file it might result in a very high bill to the S3 provider.
Additionally, it seems that many of my videos have their moov atom at the very end of the file so 20M will be just enough for a 20M video.
Given today's resolutions it doesn't cover much (changing the cap to 10M was sufficient for a 3 seconds video).

**Comment by patricktotzke** [comment]
I agree this limit should at least be configurable.
I've mounted a local directory as external file source so that's as fast as it gets yet the default settings don't generates thumbnails for videos produced by my mobile phone. Running an external conversation command on all these files. looks cumbersome..

**Comment by bkalbfus** [comment]
Please reopen this issue.  I've been living with it for a while now.  My data is on an external USB drive.  Thanks!

**Comment by br4yd** [comment]
Hello, experiencing the same issue on Nextcloud 28.0.1 using Nextcloud AIO...
Videos uploaded from an iPhone are playable without any issues. For videos uploaded from an Samsung S21 FE, a Google Pixel 2 XL or a Google Pixel 8 Pro (GrapheneOS) I get a preview image in the overview and I get sound when playing but I get no video when playing. I do get errors in the logs and that's it.

Not sure why this issue is still closed and not reopened when this issue is still existent after such a long time. Is there any known way to fix this in the current version using AIO? Unfortunately makes it impossible for me to use Nextcloud for photo and video backup when I'm not able to watch my videos.

Thanks in advance!

**Comment by br4yd** [comment]
> Your issue with not being able to play video from pixel phones is likely due to pixel phones use the x265 video codec, which is unfortunately not supported by most major browsers. There is an option to use h264 instead in the phone, but videos will take about twice as much space. The solution is to enable transcoding, for example the Memories app has the option to transcode, but it requires some extra setup and a decent CPU.

Hi, yes I do indeed have this set up and still experience the issue. I opened a related issue with more details summarized by several unanswered threads around this topic including this one. -> See here https://github.com/nextcloud/server/issues/42545

**Comment by alexanderhuzar** [comment]
All videos under 5 MB do generate previews.
Some videos larger than 5 MB do generate thumbnail previews while others don't.
Copying the videos larger than 5 MB to a Nextcloud "local" directory guarantees the thumbnails will generate.

<img width="1044" height="534" alt="Image" src="https://github.com/user-attachments/assets/1d3192b4-dfa6-4be0-9719-4a1d26d19df0" />

I'll keep mentioning this point - if you force people to become developers to solve issues like this, you'll never get enough critical mass and people will continue to use dropbox and etc. Why is everything in NextCloud so painful to do? Why do I need to go to GitHub to find answers for anything? Why do I need to ever touch the config.php or any other file when there's a GUI? Try having your parents use NextCloud if you're ever in doubt about any feature you're working on.

**Comment by nickvergessen** [comment]
> All videos under 5 MB do generate previews.

Correct

> Some videos larger than 5 MB do generate thumbnail previews while others don't.

This is because some files have the thumbnail preview stored at the beginning of the file. As per:
https://github.com/nextcloud/server/blob/bdc68989d5bbf07f9fa7a4b5a616a3c871378af3/lib/private/Preview/Movie.php#L57-L65

We are only checking the first 5MB, as we do not want to download 37G to check if there is a preview at the end of the file. The pain part here is that if we don't do this "security sanity safeguard", it might be fine for you, where you are the only user and everything is trustable,
but if a bad user has an account on your instance, they could upload big movie files and simply request previews all the time to make your Nextcloud server download 37G from your SMB share just to discard it at the end again. And then I do it again with another file and another one, keeping the connection between your SMB and your Nextcloud fully saturated, so you can no longer sync/download/upload any files.

So far for the technical / security side.

---

Now to the human side:

> I'll keep mentioning this point - if you force people to become developers to solve issues like this, you'll never get enough critical mass and people will continue to use dropbox and etc. Why is everything in NextCloud so painful to do? Why do I need to go to GitHub to find answers for anything? Why do I need to ever touch the config.php or any other file when there's a GUI? Try having your parents use NextCloud if you're ever in doubt about any feature you're working on.

🙄

**Comment by kaipee** [comment]
> > All videos under 5 MB do generate previews.
> 
> Correct
> 
> > Some videos larger than 5 MB do generate thumbnail previews while others don't.
> 
> This is because some files have the thumbnail preview stored at the beginning of the file. As per:
> 
> [server/lib/private/Preview/Movie.php](https://github.com/nextcloud/server/blob/bdc68989d5bbf07f9fa7a4b5a616a3c871378af3/lib/private/Preview/Movie.php#L57-L65)
> 
> Lines 57 to 65 in [bdc6898](/nextcloud/server/commit/bdc68989d5bbf07f9fa7a4b5a616a3c871378af3)
>  // Try downloading 5 MB first, as it's likely that the first frames are present there. 
>  // In some cases this doesn't work, for example when the moov atom is at the 
>  // end of the file, so if it fails we fall back to getting the full file. 
>  // Unless the file is not local (e.g. S3) as we do not want to download the whole (e.g. 37Gb) file 
>  if ($file->getStorage()->isLocal()) { 
>  	$sizeAttempts = [5242880, null]; 
>  } else { 
>  	$sizeAttempts = [5242880]; 
>  } 
> 
> We are only checking the first 5MB, as we do not want to download 37G to check if there is a preview at the end of the file. The pain part here is that if we don't do this "security sanity safeguard", it might be fine for you, where you are the only user and everything is trustable, but if a bad user has an account on your instance, they could upload big movie files and simply request previews all the time to make your Nextcloud server download 37G from your SMB share just to discard it at the end again. And then I do it again with another file and another one, keeping the connection between your SMB and your Nextcloud fully saturated, so you can no longer sync/download/upload any files.
> 

There are 2 issues I see here with these assumptions.

### File size

The estimation range of between 5MB and 37000MB is vast when trying to choose a sane default.

The average smartphone, for example, captures video at 1080p. With a 5MB file size, that's only a few seconds of video. Additionally 37GB is a hell of a file size for videos, even 10bit format would struggle to reach that assumed size.

More sane assumptions when trying to come to working default might be to assume:
- most modern networks operate at 1Gbps
- a low end of the scale would be a video file size of closer to 1GB - 1.5GB (assume 1080p video, at least 5 minutes in length)

### Storage type

The majority of Nextcloud instances aren't pulling from chargeable object storage (s3 and the like). It would also be a fairly simple task to implement a check for storage type, be it S3 or SMB, and set more sane file sizes based upon that.

---

TL:DR

A better, more sane, set of defaults for preview generation would be to:
- validate network storage type and limit file size based upon it
- assume most users are using 1Gbps or higher
- assume most users are mounting SMB network storage rather than cloud based object storage
- assume average user videos are personal smartphone captures at 1080p
- assume average user videos are a few minutes long (2 minutes or more)

An even more sensible approach would be to make it configurable by the server Admin, that way it can be controlled and monitored.

This should provide a much better set of defaults that will likely fix this issue for most users, and not limit generation to a file size that is barely seen since the early 2000's for video format.

**Comment by alexanderhuzar** [comment]
[Scripts.zip](https://github.com/user-attachments/files/21446521/Scripts.zip)

Ok so since NextCloud can't figure out how to cache image thumbnails, I had to do it myself thanks to a day's worth of Chat GPT giving wrong answers and crafting beautiful, but non-working scripts. Eventually I got my answer and it works great. I hope this helps someone out there. What I'm doing with ffmpeg - would be nice if NextCloud had its own thumbnail cache for movies and instead of embedding those inside the .mp4 file these would be stored in a cache directory so as to not ever touch the original file.

Basically what I'm doing now is using ffmpeg to generate a thumbnail, re-embed that thumbnail into the .mp4 file, preserve original metadata such as date taken, camera info and etc. and finally clean up with exiftool to change the filename to a standard naming convention. Cherry on top is to run the .bat script from a .vbs script so as to make it completely invisible and to not steal focus when doing something full screen - why is because I want this thing scheduled to run every hour.

Here's what I'm doing (change the directories to your own to make this work for you):

File name = "generate_thumbnails.bat"
---------------------------------------------------------------------------------------------------
@echo off
setlocal enabledelayedexpansion

REM Switch to D: drive explicitly to avoid access issues
D:

REM Directories
set "FFMPEG=D:\NextCloud\ffmpeg 2025-07-23-git-829680f96a\bin\ffmpeg.exe"
set "FFPROBE=D:\NextCloud\ffmpeg 2025-07-23-git-829680f96a\bin\ffprobe.exe"
set "TARGET=D:\NextCloud\ffmpeg 2025-07-23-git-829680f96a\test"

echo Scanning "%TARGET%" and subdirectories for MP4 files...
echo.

REM Process files recursively
for /R "%TARGET%" %%F in (*.mp4) do (
    echo Checking "%%~fF" for existing thumbnail...

    REM Check if file already has an attached thumbnail
    "%FFPROBE%" -select_streams v -show_entries stream=disposition -of default=nw=1:nk=1 "%%~fF" 2>nul | findstr /c:"1" >nul
    if not errorlevel 1 (
        echo Thumbnail already exists, skipping: "%%~fF"
        echo.
    ) else (
        echo No thumbnail detected, generating thumbnail for: "%%~fF"

        REM Generate thumbnail (frame at 1 second)
        set "THUMBFILE=%TEMP%\%%~nxF_%random%_preview.jpg"
        "%FFMPEG%" -y -i "%%~fF" -ss 00:00:01 -vframes 1 "!THUMBFILE!"

        if exist "!THUMBFILE!" (
            echo Embedding thumbnail...

            REM Output temp file
            set "TMPFILE=%TEMP%\%%~nxF_%random%_thumb.mp4"

            REM Embed thumbnail while preserving metadata
            "%FFMPEG%" -y -i "%%~fF" -i "!THUMBFILE!" -map 0 -map 1 -c copy ^
                -disposition:v:1 attached_pic -movflags faststart+use_metadata_tags -map_metadata 0 "!TMPFILE!"

            if exist "!TMPFILE!" (
                REM Save original timestamps
                for /f "tokens=*" %%A in ('powershell -NoProfile -Command "(Get-Item '%%~fF').CreationTimeUtc.ToFileTimeUtc()"') do set "CREATED=%%A"
                for /f "tokens=*" %%A in ('powershell -NoProfile -Command "(Get-Item '%%~fF').LastWriteTimeUtc.ToFileTimeUtc()"') do set "MODIFIED=%%A"
                for /f "tokens=*" %%A in ('powershell -NoProfile -Command "(Get-Item '%%~fF').LastAccessTimeUtc.ToFileTimeUtc()"') do set "ACCESSED=%%A"

                REM Replace original file
                del /F /Q "%%~fF"
                move /Y "!TMPFILE!" "%%~fF"

                REM Restore timestamps
                powershell -NoProfile -Command ^
                    "$f = Get-Item '%%~fF';" ^
                    "$f.CreationTimeUtc = [datetime]::FromFileTimeUtc(!CREATED!);" ^
                    "$f.LastWriteTimeUtc = [datetime]::FromFileTimeUtc(!MODIFIED!);" ^
                    "$f.LastAccessTimeUtc = [datetime]::FromFileTimeUtc(!ACCESSED!)"

                echo Thumbnail embedded and original file replaced: "%%~fF"
            ) else (
                echo Failed to embed thumbnail for: "%%~fF"
            )

            del /F /Q "!THUMBFILE!"
        ) else (
            echo Failed to create thumbnail for: "%%~fF"
        )
        echo.
    )
)

echo All done!
exit /b
---------------------------------------------------------------------------------------------------

File Name = thumbs and rename.vbs
---------------------------------------------------------------------------------------------------
Set WshShell = CreateObject("WScript.Shell")

' Run your first batch script (generate_thumbnails.bat)
WshShell.Run """D:\NextCloud\ffmpeg 2025-07-23-git-829680f96a\generate_thumbnails.bat""", 0, True

' Run the ExifTool command directly
WshShell.Run """D:\NextCloud\exiftool-13.31_64\exiftool.exe"" -config ""my.config"" ""-FileName<${MyDate}%+c.%e"" -d ""%Y-%m-%d %H.%M.%S"" -ignoreMinorErrors -r ""D:\NextCloud\ffmpeg 2025-07-23-git-829680f96a\test""", 0, True
---------------------------------------------------------------------------------------------------

For ExifTool the config file:
File Name = my.config
---------------------------------------------------------------------------------------------------
%Image::ExifTool::UserDefined = (
    'Image::ExifTool::Composite' => {
        MyDate => {
            Desire => {
                0 => 'CreateDate',
                1 => 'DateTimeOriginal',
                2 => 'MediaCreateDate',
                3 => 'FileModifyDate',
                4 => 'FileCreateDate',
            },
            ValueConv => q{
                # Filter out undefined or empty values from the collected dates
                my @dates = grep { $_ } @val;

                # Sort dates lexicographically (works for ISO-like formats)
                @dates = sort @dates;
                my $oldest = $dates[0] if @dates;

                # Extract only the components we need: YYYY:MM:DD HH:MM:SS
                if ($oldest and $oldest =~ /^(\d{4}):(\d{2}):(\d{2})\s+(\d{2}):(\d{2}):(\d{2})/) {
                    return sprintf("%04d-%02d-%02d %02d.%02d.%02d", $1, $2, $3, $4, $5, $6);
                }

                # Fallback: clean up the string by replacing colons if the regex doesn't match
                if ($oldest) {
                    $oldest =~ s/:/-/g;  # Replace colons with dashes
                    $oldest =~ s/\s+/ /g; # Ensure consistent spacing
                    return $oldest;
                }
            },
        },
    },
);
1;
---------------------------------------------------------------------------------------------------

**Comment by OrvilleQ** [comment]
> We are only checking the first 5MB, as we do not want to download 37G to check if there is a preview at the end of the file. The pain part here is that if we don't do this "security sanity safeguard", it might be fine for you, where you are the only user and everything is trustable,
> but if a bad user has an account on your instance, they could upload big movie files and simply request previews all the time to make your Nextcloud server download 37G from your SMB share just to discard it at the end again. And then I do it again with another file and another one, keeping the connection between your SMB and your Nextcloud fully saturated, so you can no longer sync/download/upload any files.

@nickvergessen Thank you very much for your detailed explanation.

However, as an actual user, I still hope there could be an option switch that allows downloading the entire file for preview generation - even when the storage is non-local. This would be especially useful when using preview generator. since it pre-generates all required thumbnail files in advance, eliminating the need to generate preview on fly. Therefore, once preview generator is enabled, the "bad user" scenario you described can largely be avoided.

In particular, as a Hetzner user, the traffic between my S3 bucket and my VPS which host nextcloud is completely free (as long as they are in the same data center), therefore even a bad user exist, it cannot make any damage on my wallet. But the workload of rebuild all mobile videos just to move the moov atom to the front is simply too heavy.

May I ask if nextcloud accepts AI-assisted code contributions? (Since I'm not proficient in PHP), if that's acceptable, I'd be happy to look into adding a setting that toggles whether remote storage can also fetch the entire file for preview generation. From the code snippet you shared, it doesn't seem like a very complex change.

If AI-generated contributions are not accepted, I sincerely hope you or other PHP contributors who read this could consider implementing such a feature, giving users the choice to enable or disable it according to their needs.

Thank you.

**Comment by mrsakrano** [comment]
New reply to your ticket #11443

Title: Re: [nextcloud/server] Preview of MP4 Videos failing to generate with ffmpeg/avconf if larger than 5MB (#1475)

Message: Hello,we received your message and our support will reach out to you as soon as possible.
